tools/testing/gatekeeper/skips.py

```diff
@@ -0,0 +1,102 @@
+import os
+import re
+import subprocess
+import sys
+
+import yaml
+
+
+class Checks:
+    def __init__(self):
+        config_path = os.path.join(os.path.dirname(__file__), "required-tests.yaml")
+        with open(config_path, "r", encoding="utf8") as config_fd:
+            config = yaml.load(config_fd, Loader=yaml.SafeLoader)
+        if config.get('required_tests'):
+            self.required_tests = config['required_tests']
+        else:
+            self.required_tests = []
+        if config.get('required_regexps'):
+            self.required_regexps = config['required_regexps']
+        else:
+            self.required_regexps = []
+        if config.get('paths'):
+            self.paths = OrderedDict((re.compile(key), value)
+                                       for items in config['paths']
+                                       for key, value in items.items())
+        if config.get('mapping'):
+            self.mapping = config['mapping']
+        else:
+            self.mapping = {}
+        self.all_set_of_tests = set(self.mapping.keys())
+
+    def run(self, tests, target_branch):
+        """
+        Find the required features/tests
+
+        :param: tests: report required tests+regexps (bool)
+        :param: target_branch: branch/commit to compare to
+        """
+        enabled_features = self.get_features(target_branch)
+        if not tests:
+            for feature in self.all_set_of_tests:
+                print(f"skip_{feature}=" +
+                      ('no' if feature in enabled_features else 'yes'))
+            return 0
+        required_tests = set(self.required_tests)
+        required_regexps = set(self.required_regexps)
+        for feature in enabled_features:
+            values = self.mapping.get(feature, {})
+            if values.get("names"):
+                required_tests.update(values["names"])
+            if values.get("regexps"):
+                required_regexps.add(values["regexps"])
+        print(';'.join(required_tests))
+        print('|'.join(required_regexps))
+        return 0
+
+    def get_features(self, target_branch):
+        """
+        Get changed file to `target_branch` and map them to the
+        to-betested set-of-tests
+
+        :param target_branch: branch/commit to compare to
+        :returns: List of set-of-tests
+        """
+        cmd = ["git", "diff", "--name-only", f"origin/{target_branch}"]
+        changes = [_.decode("utf-8")
+                   for _ in subprocess.check_output(cmd).split(b'\n')
+                   if _.strip()]
+        print('\n'.join(changes), file=sys.stderr)
+        enabled_features = set()
+        # Go through lines and find what features should be covered
+        for line in changes:
+            for regexp, features in self.paths.items():
+                if regexp.search(line):
+                    for feature in features:
+                        enabled_features.add(feature)
+                    break
```

Comment:
Hi @ldoktor !

This `break` is very important, as it applies the `sort by order of importance` rule. For example,  if the first match suggest an empty list of features then, in practice, it is a skip; if there are other matches to this same line that suggest a list of features then they won't apply. So shall we get this line document to avoid future bugs by accident removal of this `break` or even a change on the algorithm?

Reply:
The same here, I'll add a comment in v2 (my code/comment ratio is way below my average as this is a product of many failed attempts)